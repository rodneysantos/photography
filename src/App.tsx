import {
	Component,
	Suspense,
	createResource,
	createSignal,
	onMount,
} from "solid-js";
import { Gallery, Header, Lightbox, MenuKey } from "./components";
import type { Photo } from "./models";
import { fetchPhotosByTags } from "./service";
import {
	getSelectedPhoto,
	hasSelectedPhoto,
	updateSelectedPhoto,
} from "./service/url";

const App: Component = () => {
	const [clickedPhoto, setClickedPhoto] = createSignal<Photo>();
	const [activeMenuItems, setActiveMenuItems] = createSignal<MenuKey[]>([
		"bnw",
		"low-key",
	]);
	const [photos, { refetch }] = createResource<Photo[], MenuKey[]>(
		activeMenuItems,
		fetchPhotosByTags,
		{ initialValue: [] },
	);

	onMount(() => {
		if (hasSelectedPhoto()) {
			setClickedPhoto({ url: getSelectedPhoto()! });
		}
	});

	function handleActiveItemsChange(activeItems: MenuKey[]) {
		setActiveMenuItems(activeItems);
		refetch();
	}

	function handlePhotoClick(photo: Photo) {
		setClickedPhoto(photo);
		updateSelectedPhoto(photo);
	}

	function handleClose() {
		setClickedPhoto(undefined);
		updateSelectedPhoto(null);
	}

	return (
		<div data-testid="app" class="flex flex-col">
			<Header onActiveItemsChange={handleActiveItemsChange} />

			<Suspense fallback={<p>Loading...</p>}>
				<Gallery photos={photos()} onPhotoClick={handlePhotoClick} />
			</Suspense>

			<Lightbox photo={clickedPhoto()} onClose={handleClose} />
		</div>
	);
};

export default App;
