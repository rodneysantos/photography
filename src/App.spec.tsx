import { cleanup, render } from "@solidjs/testing-library";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import App from "./App";

describe("App", () => {
	beforeEach(() => {
		vi.useFakeTimers();
		vi.mock("./service", () => ({
			fetchPhotosByTags: () => Promise.resolve(["photo"]),
		}));
	});

	afterEach(() => {
		vi.clearAllTimers();
	});

	afterEach(cleanup);

	it("renders correctly", () => {
		// arrange
		const { asFragment } = render(() => <App />);

		// assert
		expect(asFragment()).toMatchSnapshot();
	});

	it("updates the gallery on menu item click", async () => {
		// arrange
		const { getByTestId } = render(() => <App />);

		// act
		getByTestId("color").click();
		await vi.runAllTimersAsync();

		// assert
		expect(getByTestId("gallery-list").children.length).toBe(1);
	});
});
