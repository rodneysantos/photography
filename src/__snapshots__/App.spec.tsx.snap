// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders correctly 1`] = `"<div data-testid=\\"app\\" class=\\"flex flex-col\\"><div class=\\"flex flex-col text-center mt-14 5xl:mt-16 5xl:mb-4\\"><div class=\\"z-10 mb-4 bg-white\\"><div class=\\"tracking-widest text-3xl font-big-shoulder\\">RODNEY SANTOS</div><div class=\\"text-xl tracking-xs translate-x-1/10 font-dosis\\">PHOTOGRAPHY</div></div><div data-testid=\\"menuContainer\\" class=\\"flex flex-col items-center transition-all will-change-transform -translate-y-14\\"><div data-testid=\\"menu\\" class=\\"transition-opacity will-change-auto opacity-0\\"><div class=\\"flex flex-row justify-center\\"><div class=\\"px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 bg-neutral-900 text-neutral-100\\" data-testid=\\"bnw\\">Black and White</div><div class=\\"px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900\\" data-testid=\\"color\\">Color</div><div class=\\"px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 bg-neutral-900 text-neutral-100\\" data-testid=\\"low-key\\">Low-Key</div></div></div><svg data-testid=\\"chevronDownIcon\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\" class=\\"w-5 mt-4 transition-opacity cursor-pointer will-change-auto\\"><path d=\\"M256 0a256 256 0 1 0 0 512A256 256 0 1 0 256 0zM135 241c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0l87 87 87-87c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9L273 345c-9.4 9.4-24.6 9.4-33.9 0L135 241z\\"></path></svg></div></div><p>Loading...</p></div>"`;
